cmd/gaia/app/app.go

```diff
@@ -169,32 +176,150 @@ func NewGenesisAccount(acc *auth.BaseAccount) GenesisAccount {
+		chainID = cmn.Fmt("test-chain-%v", cmn.RandStr(6))
+	}
+
+	// get genesis flag account information
+	accountsStr := viper.GetString(flagAccounts)
+	accounts := strings.Split(accountsStr, ":")
+	genaccs := make([]GenesisAccount, len(accounts))
+	for i, account := range accounts {
+		p := strings.Split(account, "-")
+		if len(p) != 3 {
+			err = errors.New("input account has bad form, each account must be in form name-coins-isval, for example: foobar-10fermion,10baz-true")
+			return
+		}
+		name := p[0]
+		var coins sdk.Coins
+		coins, err = sdk.ParseCoins(p[1])
+		if err != nil {
+			return
+		}
+		isValidator := false
+		if p[2] == "true" {
+			isValidator = true
+		}
+
+		var addr sdk.Address
+		var secret string
+		addr, secret, err = server.GenerateCoinKey()
+		if err != nil {
+			return
+		}
+
+		printMap["secret-"+name] = secret
+
+		// create the genesis account
+		accAuth := auth.NewBaseAccountWithAddress(addr)
+		accAuth.Coins = coins
+		acc := NewGenesisAccount(&accAuth)
+		genaccs[i] = acc
+
+		// add the validator
+		if isValidator {
+
+			// only use this machines pubkey the first time, all others are dummies
+			var pk crypto.PubKey
+			if i == 0 {
+				pk = pubKey
+			} else {
+				pk = crypto.GenPrivKeyEd25519().PubKey()
+			}
+
+			freePower := int64(100)
+			validator := tmtypes.GenesisValidator{
+				PubKey: pk,
+				Power:  freePower,
+			}
+			desc := stake.NewDescription(name, "", "", "")
+			candidate := stake.NewCandidate(addr, pk, desc)
+			candidate.Assets = sdk.NewRat(freePower)
+			poolAssets += freePower
+			validators = append(validators, validator)
+			candidates = append(candidates, candidate)
+		}
+	}
+
+	// create the print message
+	bz, err := cdc.MarshalJSON(printMap)
+	cliPrint = json.RawMessage(bz)
+
+	stakeData := stake.GetDefaultGenesisState()
+	stakeData.Candidates = candidates
+
+	// assume everything is bonded from the get-go
+	stakeData.Pool.TotalSupply = poolAssets
+	stakeData.Pool.BondedPool = poolAssets
+	stakeData.Pool.BondedShares = sdk.NewRat(poolAssets)
 
 	genesisState := GenesisState{
 		Accounts:  genaccs,
-		StakeData: stake.GetDefaultGenesisState(),
+		StakeData: stakeData,
 	}
 
-	stateBytes, err := json.MarshalIndent(genesisState, "", "\t")
+	appState, err = wire.MarshalJSONIndent(cdc, genesisState)
+	return
+}
+
+// append gaia app_state together, stitch the accounts together take the
+// staking parameters from the first appState
+func GaiaAppendAppState(cdc *wire.Codec, appState1, appState2 json.RawMessage) (appState json.RawMessage, err error) {
+	var genState1, genState2 GenesisState
+	err = cdc.UnmarshalJSON(appState1, &genState1)
 	if err != nil {
-		return nil, err
+		panic(err)
 	}
-
-	return stateBytes, nil
+	err = cdc.UnmarshalJSON(appState2, &genState2)
+	if err != nil {
+		panic(err)
+	}
+	genState1.Accounts = append(genState1.Accounts, genState2.Accounts...)
+	genState1.StakeData.Candidates = append(genState1.StakeData.Candidates, genState2.StakeData.Candidates...)
```

Comment:
How to add the subfield of `GenesisState` for staking could be specified instead in the `staking` module, so other applications could reuse it.

Reply:
I agree that this should be refactored to allow for reuse from the stake module